File: creational/builder-example.py
from enum import Enum


class DoorNumber(Enum):
    UNDEFINED = -1
    TWO = 0
    FOUR = 1


class RoomSize(Enum):
    UNDEFINED = -1
    SMALL = 0
    MEDIUM = 1
    LARGE = 2


class Car:
    def __init__(self):
        self.__room_size = RoomSize.UNDEFINED
        self.__door_number = DoorNumber.UNDEFINED
        self.__build_finished = False

    def show_car_info(self):
        print('Room Size:', self.__room_size)
        print('Door Number:', self.__door_number)

    @staticmethod
    def build_general(room_size, door_number):
        if not isinstance(room_size, RoomSize) or not isinstance(door_number, DoorNumber):
            raise Exception('Wrong type of parameters are passed!')

        car = Car()
        car.room_size = room_size
        car.door_number = door_number
        return car

    @staticmethod
    def build_small_car():
        car = Car()
        car.room_size = RoomSize.SMALL
        car.door_number = DoorNumber.TWO
        return car

    def set_room_size(self, room_size):
        self.__room_size = room_size
        return self

    def set_door_number(self, door_number):
        self.__door_number = door_number
        return self

    @staticmethod
    def builder():
        return Car()

    def build(self):
        self.__build_finished = True
        return self


if __name__ == "__main__":
    c1 = Car.build_general(RoomSize.LARGE, DoorNumber.FOUR)
    c1.show_car_info()

    print('- - - - - - - - - - - - - - - - -')
    c2 = Car.build_small_car()
    c2.show_car_info()

    print('- - - - - - - - - - - - - - - - -')
    c3 = Car.builder().set_room_size(RoomSize.LARGE).set_door_number(DoorNumber.FOUR).build()
    c3.show_car_info()

File: creational/builder-example.js
class DoorNumber { }
DoorNumber.UNDEFINED = -1
DoorNumber.TWO = 0
DoorNumber.FOUR = 1


class RoomSize { }
RoomSize.UNDEFINED = -1
RoomSize.SMALL = 0
RoomSize.MEDIUM = 1
RoomSize.LARGE = 2

class Car {
    constructor() {
        this.roomSize = RoomSize.UNDEFINED;
        this.doorNumber = DoorNumber.UNDEFINED;
        this.buildFinished = false;
    }

    showCarInfo() {
        console.log('Room Size:', this.roomSize);
        console.log('Door Number:', this.doorNumber);
    }

    static buildGeneral(room_size, door_number) {
        if (!(Object.keys(RoomSize).hasOwnProperty(room_size)) || !(Object.keys(DoorNumber).hasOwnProperty(door_number))) {
            throw new Error('Wrong type of parameters are passed!');
        }

        const car = new Car();
        car.roomSize = room_size;
        car.doorNumber = door_number;
        return car;
    }

    static buildSmallCar() {
        const car = new Car();
        car.roomSize = RoomSize.SMALL;
        car.doorNumber = DoorNumber.TWO;
        return car;
    }

    set_roomSize(room_size) {
        this.roomSize = room_size;
        return this;
    }

    set_doorNumber(door_number) {
        this.doorNumber = door_number;
        return this;
    }

    static builder() {
        return new Car();
    }

    build() {
        this.buildFinished = true;
        return this;
    }
}


const c1 = Car.buildGeneral(RoomSize.LARGE, DoorNumber.FOUR);
c1.showCarInfo();

console.log('- - - - - - - - - - - - - - - - -');
const c2 = Car.buildSmallCar();
c2.showCarInfo();

console.log('- - - - - - - - - - - - - - - - -');
const c3 = Car.builder().set_roomSize(RoomSize.LARGE).set_doorNumber(DoorNumber.FOUR).build();
c3.showCarInfo();